### PR TITLE
Add task info in exporter

### PR DESCRIPTION
**Pull request by ninjadq**

### CODE REVIEW

**Comment by ninjadq** [comment]
image

File: src/pkg/exporter/js_collector.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+// JobServiceCollectorName ...
+const JobServiceCollectorName = "JobServiceCollector"
+
+var (
+	jobServiceTaskQueueSize = typedDesc{
+		desc:      newDescWithLables("", "task_queue_size", "Total number of tasks", "type"),
+		valueType: prometheus.GaugeValue,
+	}
+	jobServiceTaskQueueLatency = typedDesc{
+		desc:      newDescWithLables("", "task_queue_latency", "the time last taks processed", "type"),
```
Comment (bitsf) [inline src/pkg/exporter/js_collector.go:17]: typo

File: src/pkg/exporter/js_collector.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		valueType: prometheus.GaugeValue,
+	}
+	jobServiceConcurrency = typedDesc{
+		desc:      newDescWithLables("", "task_concurrecy", "Total number of concurrency on a pool", "type", "pool"),
```
Comment (bitsf) [inline src/pkg/exporter/js_collector.go:21]: typo

File: src/pkg/exporter/cache.go [hunk 3]
```diff
@@ -86,11 +86,8 @@ func CacheInit(opt *Opt) {
 			cacheCleanInterval = defaultCacheCleanInterval
 		}
 		ticker := time.NewTicker(time.Duration(cacheCleanInterval) * time.Second)
-		for {
-			select {
-			case <-ticker.C:
-				StartCacheCleaner()
-			}
+		for range ticker.C {
```
Comment (steven-zou) [inline src/pkg/exporter/cache.go:89]: no defer close func for ticker?

File: src/pkg/exporter/js_worker.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	})
+	if err != nil {
+		panic(err)
+	}
+	redisPool = pool
+}
+
+// InitBackendWorker initiate backend worker
+func InitBackendWorker(redisPoolConfig *RedisPoolConfig) {
+	pool, err := redislib.GetRedisPool("JobService", redisPoolConfig.URL, &redislib.PoolParam{
```
Comment (steven-zou) [inline src/pkg/exporter/js_worker.go:49]: Why create two Redis pools?

File: src/pkg/exporter/js_collector.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	result := []prometheus.Metric{}
+
+	// Get concurrency info via raw redis client
+	rdsConn := GetRedisPool().Get()
+	values, err := redis.Values(rdsConn.Do("SMEMBERS", redisKeyKnownJobs(jsNamespace)))
+	checkErr(err, "err when get known jobs")
+	var jobs []string
+	for _, v := range values {
+		jobs = append(jobs, string(v.([]byte)))
```
Comment (steven-zou) [inline src/pkg/exporter/js_collector.go:74]: Why not put the logic in one `for`?

Comment (steven-zou) [reply src/pkg/exporter/js_collector.go:74]: With `for` in line 76

File: src/pkg/exporter/js_collector.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	for _, job := range jobs {
+		values, err := redis.Values(rdsConn.Do("HGETALL", redisKeyJobsLockInfo(jsNamespace, job)))
+		checkErr(err, "err when get job lock info")
+		for i := 0; i < len(values); i += 2 {
+			key, _ := redis.String(values[i], nil)
+			value, _ := redis.Float64(values[i+1], nil)
+			result = append(result, jobServiceConcurrency.MustNewConstMetric(value, job, key))
+		}
+	}
+	rdsConn.Close()
```
Comment (steven-zou) [inline src/pkg/exporter/js_collector.go:85]: defer close may be more clear?

File: src/pkg/exporter/js_worker.go (hunk 4 above)
Comment (steven-zou) [reply src/pkg/exporter/js_worker.go:49]: With line 34

**Review by steven-zou — COMMENTED** [review]
See my comments?

File: src/pkg/exporter/js_worker.go (hunk 4 above)
Comment (ninjadq) [reply src/pkg/exporter/js_worker.go:49]: one for redis connection. The other one is reuse of `backednWorker`, so it didn't to implements the same logic on exporter.

**Review by bitsf — APPROVED** [review]

File: src/pkg/exporter/cache.go (hunk 3 above)
Comment (steven-zou) [reply src/pkg/exporter/cache.go:89]:
```go
defer ticker.Close()
```

???

File: src/pkg/exporter/js_collector.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// JobServiceCollector ...
+type JobServiceCollector struct {
+	Namespace string
+}
+
+// Describe implements prometheus.Collector
+func (hc *JobServiceCollector) Describe(c chan<- *prometheus.Desc) {
+	c <- jobServiceTaskQueueSize.Desc()
```
Comment (steven-zou) [inline src/pkg/exporter/js_collector.go:46]: Are all the `Desc()` returning `typedDesc`? If yes, why not make the logic simpler?

```go
d:=[]typedDesc{
  jobServiceTaskQueueSize.Desc(),
  jobServiceTaskQueueLatency.Desc(),
  jobServiceConcurrency.Desc(),
  jobServiceScheduledJobFails.Desc(),
}

for _, jd := range d {
   c <- jd
}

```

File: src/pkg/exporter/js_collector.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+		if ok {
+			return value.([]prometheus.Metric)
+		}
+	}
+	result := []prometheus.Metric{}
+
+	// Get concurrency info via raw redis client
+	rdsConn := GetRedisPool().Get()
+	defer rdsConn.Close()
+	knownJobvalues, err := redis.Values(rdsConn.Do("SMEMBERS", redisKeyKnownJobs(jsNamespace)))
```
Comment (steven-zou) [inline src/pkg/exporter/js_collector.go:71]: If possible, I suggest you can put the Redis read/write logic into a separate func for clarification.

File: src/pkg/exporter/js_worker.go (hunk 4 above)
Comment (steven-zou) [reply src/pkg/exporter/js_worker.go:49]: As it's a ` pool`, I think one is enough if it is used in the same process.

Comment (steven-zou) [reply src/pkg/exporter/js_worker.go:49]: As an example, the job service only has one Redis pool.

**Review by steven-zou — COMMENTED** [review]
More comments, please take a look.

**Review by steven-zou — APPROVED** [review]
LGTM

**Pull request merged**